/dev/vc-mem read() using DMA

hvenev opened the pull request:
I've updated https://git.venev.name/hristo/rpi-vc-log to use `pread()` and it works on my Pi 4.

Note that we rely on the fact that the DMA channel has the same view of memory as the VPU. Using the 40-bit DMA channels will require translating addresses somehow.

The way we map VPU addresses to ARM physical addresses needs some work. Instead of just adding some `phys_addr`, we should probably go through the device tree `ranges` for every page, and fail to map the ones that cannot be translated.

Edit: Can you please explain why something is mangling `/axi/ranges`?

hvenev committed:
bcm2835-dma: Drop bcm2711_dma40_memcpy.

Even though this function was exported, it was not defined in any of the
kernel headers, and it has no in-tree users.

Furthermore, by dropping it we free up one 40-bit DMA channel.

Signed-off-by: Hristo Venev <hristo@venev.name>

hvenev committed:
dts: bcm2835-rpi: Add vc_mem node.

Firmware already sets `/axi/reg`, so put the device tree node there.

Signed-off-by: Hristo Venev <hristo@venev.name>

hvenev committed:
vc-mem: Drop some unused/empty functions.

vc_mem_get_current_size does not have any in-tree users, so it can be
dropped. Also remove two empty static functions.

Signed-off-by: Hristo Venev <hristo@venev.name>

hvenev committed:
vc-mem: Convert into a platform device.

The compatible string of the device is raspberrypi,vc-mem. The user API is
mostly the same.

TODO: Correctly map VPU-side addresses to ARM addresses in mmap().

Signed-off-by: Hristo Venev <hristo@venev.name>

hvenev committed:
vc-mem: Implement reading using DMA.

read() and pread() now use DMA to read addresses that the ARM core cannot
directly access.

Signed-off-by: Hristo Venev <hristo@venev.name>

hvenev committed:
bcm2835-dma: Free up another channel.

Don't allocate a channel for the legacy API if it is disabled.

Signed-off-by: Hristo Venev <hristo@venev.name>

hvenev commented:
It looks like this also works on Pi 3B+.

pelwell commented:
That's good to hear. I wasn't to take a while to understand what you've done and why, but I'll give you some feedback in the next few days.

hvenev committed:
dt-bindings: firmware: Document raspberrypi,vc-mem

The raspberrypi,vc-mem node provides access to the VideoCore, which is used
by various debugging tools.

Signed-off-by: Hristo Venev <hristo@venev.name>

hvenev commented:
Documented dt binding and rebased.
